README.md
# banking_code_challenge
Code challenge for Digital Origin

Install
-------

* First execute 'bundle install' to install gems

* Run 'rake' to execute features in show_me_the_money.rb file.

* Run 'rake rspec' to run all tests.


Description
-----------

## About​ ​This​ ​Test

The purpose of this code test is to show us your skills in:

* Object Oriented Programming
* TDD
* Clean code
* Git
* Problem solving

Please keep these aspects in mind as you develop your solution. Also, your
chosen implementation doesn't necessarily have to be the best you can think of,
but one that you can implement in the allocated time.

## Instructions

* Create a project which models the situation below.
* The language used should be Ruby​ (unless previously agreed otherwise).
* All code should be tested following TDD​.
* A README​ file must be attached explaining your implementation.
* Create a .zip with your code​ and git​ directory.
* Answer our email with the zip​ of the project.

## Background

The software you write in this test will be used for banks. Banks have accounts.
Accounts hold money. Transfers can be made between accounts. Banks store the
history of transfers.

There can be two types of transfers:

* Intra-bank​ transfers, between accounts of the same bank. They don't
have commissions, they don't have limits and they always succeed.
* Inter-bank​ transfers, between accounts of different banks. They have 5€
commissions, they have a limit of 1000€ per transfer. And they have a 30%
chance of failure.


Features
--------

## Part​ ​1

Create the models and their relationships to accurately reflect banks, accounts
and transfers. Make sure that new types of accounts and transfers can be added
to the bank with minimal effort.

## Part​ ​2

After modeling the domain(part 1), create a file called
show_me_the_money.rb*​ in charge of reproducing the next situation:

Jim has an account on the bank A and Emma has an account on the bank B. Jim
owes Emma 20000€. Emma is already a bit angry, because she did not get the
money although Jim told her that he already sent it.

Help Jim send his money by developing a transfer​ ​agent​. This agent assures
that everybody gets their money. When the agent receives an order to transfer
money from account A to account B, he issues transfers considering commissions,
transfer limits and possibility of transfer failures.

The execution of the script will print the balance of every account before and after
the transfers and the history of the transfer of every bank.

Questions:
----------

Please also supply your answer to the following questions:

##  How would you improve your solution?

- If i'll know all the characteristics and purposes of the project maybe i would change parts of the models, adding a model for holders accounts.

- Also i think i could have added more security to transfers, like add a password to accounts or something to make sure the account's holder previously approved transfer.

- Also i could handle more accurate the part of making fail the 30% of Transfers randomly, i think it would be better if it depended on something more clear.

##  How would you adapt your solution if transfers are not instantaneous?

- Using Callbacks inside my Processes or a solution to handle the queuing system, like Sidekiq. This prevents the user from waiting and allows to control transfers errors to avoid unwanted deliveries.

Requirements
------------

* ruby-2.2


Author
------

Original author: Manuel Figueroa Andrade


License
-------

(The MIT License) FIXME (different license?)

Copyright (c) 2018 FIXME (author's name)

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.txt
== 1.0.0 / 2018-01-14

* 1 major enhancement
  * Birthday!


spec/bank_spec.rb

require File.expand_path('../spec_helper', __FILE__)
Dir["./lib/*.rb"].each {|file| require File.expand_path('../../'+file, __FILE__)}

describe BankingCodeChallenge::Bank do

  before :each do
		@bank1 = BankingCodeChallenge::Bank.new( "Bank1" )
    @account_1_1 = BankingCodeChallenge::Account.new(1111,@bank1,1000)
    @account_1_2 = BankingCodeChallenge::Account.new(1122,@bank1,2000)
    @transfer1 = BankingCodeChallenge::Transfer.new(@account_1_1, @account_1_2, 500)
    @bank2 = BankingCodeChallenge::Bank.new( "Bank2" )

	end

  describe 'validations' do

		context '.initialize' do
			it '#bank' do
        expect(@bank1).to be_an_instance_of(BankingCodeChallenge::Bank)
			end
    end

    context '.addAccount' do
			it 'successfully' do
        @bank1.addAccount(@account_1_2)
        expect(@bank1.accounts).to include(@account_1_2)
			end

      it 'with repeated number' do
        @bank1.addAccount(@account_1_2)
        expect{@bank1.addAccount(@account_1_2)}.to raise_error(RuntimeError)
			end

	  end

  end

end


spec/agent_spec.rb

require File.expand_path('../spec_helper', __FILE__)
Dir["./lib/*.rb"].each {|file| require File.expand_path('../../'+file, __FILE__)}

describe BankingCodeChallenge::Agent do

  before :each do
    @bank1 = BankingCodeChallenge::Bank.new( "Bank1" )
    @bank2 = BankingCodeChallenge::Bank.new( "Bank2" )
    @account_1_1 = BankingCodeChallenge::Account.new(1111,@bank1,100000)
    @account_2_1 = BankingCodeChallenge::Account.new(1111,@bank2,3000)
    @inter_bank_1 = BankingCodeChallenge::InterBank.new(@account_1_1, @account_2_1, 100)
    @inter_bank_2 = BankingCodeChallenge::InterBank.new(@account_1_1, @account_2_1, 100000)
    @agent = BankingCodeChallenge::Agent.new()

	end

  describe 'validations' do

		context '.initialize' do
			it '#agent' do
        expect(@agent).to be_an_instance_of(BankingCodeChallenge::Agent)
			end
    end

    context '#agent' do
			it '.makeTransfer successfully' do
        expect{@agent.makeTransfer(20000,@account_1_1,@account_2_1)}.to change{@account_1_1.balance}.by(-20105)
        expect{@agent.makeTransfer(20000,@account_1_1,@account_2_1)}.to change{@account_2_1.balance}.by(20000)
        expect(@agent.makeTransfer(20000,@account_1_1,@account_2_1)).to eq ('successfully delivered')
			end

      it '.makeTransfer fail' do
        expect{@agent.makeTransfer(200000,@account_1_1,@account_2_1)}.to raise_error(RuntimeError)
			end

      it '.checkTotalCommissions successfully' do
        expect(@agent.checkTotalCommissions @inter_bank_1).to have_key(:maxAmount)
			end

      it '.checkTotalCommissions fail' do
        expect{@agent.checkTotalCommissions @inter_bank_2}.to raise_error(RuntimeError)
			end

      it '.deliver' do
        expect(@agent.deliver(@inter_bank_1)).to eq ('successfully delivered')
      end

	  end

  end

end


spec/spec_helper.rb

require 'rubygems'
require 'rspec'

require File.expand_path('../../lib/banking_code_challenge', __FILE__)

RSpec.configure do |config|
  # == Mock Framework
  #
  # RSpec uses it's own mocking framework by default. If you prefer to
  # use mocha, flexmock or RR, uncomment the appropriate line:
  #
  # config.mock_framework = :mocha
  # config.mock_framework = :flexmock
  # config.mock_framework = :rr
end



lib/agent.rb
require 'colorize'
require 'intra_bank.rb'
require 'inter_bank.rb'

module BankingCodeChallenge

  class Agent

    def makeTransfer amount, senderAccount, beneficiaryAccount

      # Intra-bank transfers
      if beneficiaryAccount.bank == senderAccount.bank
        raise "ALERT! Not enough Money in sender account, try a smaller amount.".red if senderAccount.balance < amount
        raise "ALERT! Sender and beneficiary most be different.".red if senderAccount == beneficiaryAccount

        intra_bank = IntraBank.new(senderAccount, beneficiaryAccount, amount)
        intra_bank.deliver
        intra_bank.status

      # Inter-bank transfers
      else
        @inter_bank = InterBank.new(senderAccount, beneficiaryAccount, amount)

        raise "ALERT! Not enough Money in sender account, try a smaller amount.".red if senderAccount.balance < (amount + @inter_bank.commission)

        if amount > @inter_bank.limit
          # Split transfer to avoid exceeds limit
          splitTransfers = self.checkTotalCommissions @inter_bank

          status = self.handleSplitTransfers splitTransfers

        else

          status = self.deliver @inter_bank

        end

        status

      end
    end

    def checkTotalCommissions inter_bank

      counter = 0
      maxAmount = inter_bank.limit - inter_bank.commission
      accumulated = maxAmount

      while accumulated < inter_bank.amount
        accumulated = accumulated + maxAmount
        counter += 1
      end
      lastTransfer = inter_bank.amount - maxAmount*(counter)
      # cost = (maxAmount+@inter_bank.commission)*counter + lastTransfer + @inter_bank.commission
      totalCommissions = inter_bank.commission*(counter+1)

      if inter_bank.amount + totalCommissions <= inter_bank.senderAccount.balance
        {maxAmount: maxAmount, maxAmountTimes: counter, lastTransfer: lastTransfer}
      else
        raise "ALERT! Not enough Money in sender account, try a smaller amount.".red
      end

    end

    def handleSplitTransfers splitTransfers

      splitTransfer = InterBank.new(@inter_bank.senderAccount, @inter_bank.beneficiaryAccount, splitTransfers[:maxAmount])
      lastTransfer = InterBank.new(@inter_bank.senderAccount, @inter_bank.beneficiaryAccount, splitTransfers[:lastTransfer])

      (splitTransfers[:maxAmountTimes]).times do
        self.deliver splitTransfer
      end

      self.deliver lastTransfer

    end

    def deliver inter_bank

      inter_bank.deliver

      while inter_bank.status == 'failed delivery'
        inter_bank.deliver
      end

      inter_bank.status
    end

  end

end


lib/inter_bank.rb
require 'transfer.rb'

module BankingCodeChallenge

  class InterBank < Transfer

    attr_reader :senderAccount, :beneficiaryAccount, :amount, :commission, :limit, :failure, :status

    def initialize senderAccount, beneficiaryAccount, amount, commission = 5, limit = 1000, failure = 30
      super(senderAccount, beneficiaryAccount, amount)
      @commission = commission
      @limit = limit
      @failure = failure
      @status = 'pending'
    end

    def deliver

      if self.senderAccount.balance < self.amount || Random.rand(1..10) <= 3
        @status = 'failed delivery'
      else
        self.senderAccount.takeOutMoney(self.amount+self.commission)
        self.beneficiaryAccount.depositMoney(self.amount)
        @status = 'successfully delivered'
      end

    end

    def updateAmount amount
      @amount = amount
    end

  end

end


lib/banking_code_challenge.rb
require 'pry'

module BankingCodeChallenge

  # :stopdoc:
  LIBPATH = ::File.expand_path('..', __FILE__) + ::File::SEPARATOR
  PATH = ::File.dirname(LIBPATH) + ::File::SEPARATOR
  VERSION = ::File.read(PATH + 'version.txt').strip
  # :startdoc:

  # Returns the library path for the module. If any arguments are given,
  # they will be joined to the end of the libray path using
  # <tt>File.join</tt>.
  #
  def self.libpath( *args )
    rv =  args.empty? ? LIBPATH : ::File.join(LIBPATH, args.flatten)
    if block_given?
      begin
        $LOAD_PATH.unshift LIBPATH
        rv = yield
      ensure
        $LOAD_PATH.shift
      end
    end
    return rv
  end

  # Returns the lpath for the module. If any arguments are given,
  # they will be joined to the end of the path using
  # <tt>File.join</tt>.
  #
  def self.path( *args )
    rv = args.empty? ? PATH : ::File.join(PATH, args.flatten)
    if block_given?
      begin
        $LOAD_PATH.unshift PATH
        rv = yield
      ensure
        $LOAD_PATH.shift
      end
    end
    return rv
  end

  # Utility method used to require all files ending in .rb that lie in the
  # directory below this file that has the same name as the filename passed
  # in. Optionally, a specific _directory_ name can be passed in such that
  # the _filename_ does not have to be equivalent to the directory.
  #
  def self.require_all_libs_relative_to( fname, dir = nil )
    dir ||= ::File.basename(fname, '.*')
    search_me = ::File.expand_path(
        ::File.join(::File.dirname(fname), dir, '**', '*.rb'))

    Dir.glob(search_me).sort.each {|rb| require rb}
  end

end  # module BankingCodeChallenge

BankingCodeChallenge.require_all_libs_relative_to(__FILE__)


lib/show_me_the_money.rb
require 'rspec'
require 'pry'
require 'colorize'


puts "\nBANKING CODE CHALLENGE".yellow
puts "XXXXXXXXXXXXXXXXXXXXXX".yellow

puts "\nPART 1".magenta
puts "\nCreate the models and their relationships to accurately reflect banks, accounts and transfers.".green

bank1 = BankingCodeChallenge::Bank.new("Bank1")
account_1_1 = BankingCodeChallenge::Account.new(1111,bank1,50500, "Jim")
account_1_2 = BankingCodeChallenge::Account.new(1122,bank1,2000)
bank1.addAccount(account_1_1)
bank1.addAccount(account_1_2)

puts "\nFirst bank name: " + bank1.name
puts "\nFirst bank accounts: "
bank1.accounts.map{|a|
  puts "account number: " + a.number.to_s + " balance: €" + a.balance.to_s
}

bank2 = BankingCodeChallenge::Bank.new("Bank2")
account_2_1 = BankingCodeChallenge::Account.new(2211,bank2,3000, "Emma")
account_2_2 = BankingCodeChallenge::Account.new(2222,bank2,4000)
bank2.addAccount(account_2_1)
bank2.addAccount(account_2_2)

puts "\nSecond bank name: " + bank2.name
puts "\nSecond bank accounts: "
bank2.accounts.map{|a|
  puts "account number: " + a.number.to_s + " balance: €" + a.balance.to_s
}

puts "\nTransfer €500 by INTRA-BANK transfer".green
puts "\nBEFORE transfer".green
puts bank1.name
puts "account number: " + account_1_1.number.to_s + " balance: €" + account_1_1.balance.to_s
puts "account number: " + account_1_2.number.to_s + " balance: €" + account_1_2.balance.to_s
puts "history of the transfers: " + bank1.transfers.count.to_s

intraBank = BankingCodeChallenge::IntraBank.new(account_1_1, account_1_2, 500)
intraBank.deliver
puts "\nTransfering...".cyan

puts "\nAFTER transfer".green
puts bank1.name
puts "Transfer status: " + intraBank.status
puts "account number: " + account_1_1.number.to_s + " balance: €" + account_1_1.balance.to_s
puts "account number: " + account_1_2.number.to_s + " balance: €" + account_1_2.balance.to_s
puts "history of the transfers: " + bank1.transfers.count.to_s


puts "\nTransfer €500 by INTER-BANK transfer".green
puts "\nBEFORE transfer".green
puts bank1.name
puts "account number: " + account_1_1.number.to_s + " balance: €" + account_1_1.balance.to_s
puts "history of the transfers: " + bank1.transfers.count.to_s

puts bank2.name
puts "account number: " + account_2_1.number.to_s + " balance: €" + account_2_1.balance.to_s
puts "history of the transfers: " + bank2.transfers.count.to_s


interBank = BankingCodeChallenge::InterBank.new(account_1_1, account_2_1, 500)
interBank.deliver
puts "\nTransfering...".cyan

puts "\nAFTER transfer".green
if interBank.status == "failed delivery"
  puts "Transfer status: Transfer failed by 30% chance of failure."
else
  puts "Transfer status: " + interBank.status
end
puts bank1.name
puts "account number: " + account_1_1.number.to_s + " balance: €" + account_1_1.balance.to_s
puts "history of the transfers: " + bank1.transfers.count.to_s
puts bank2.name
puts "account number: " + account_2_1.number.to_s + " balance: €" + account_2_1.balance.to_s
puts "history of the transfers: " + bank2.transfers.count.to_s



puts "\nPART 2".magenta

puts "\nJim has an account on the bank A and Emma has an account on the bank B."
puts "Jim owes Emma 20000€. Emma is already a bit angry, because she did not get the money although Jim told her that he already sent it."
puts "Help Jim send his money by developing a transfer​ ​agent​."

agent = BankingCodeChallenge::Agent.new()

puts "This agent assures that everybody gets their money."
puts "When the agent receives an order to transfer money from account A to account B,"
puts "he issues transfers considering commissions,transfer limits and possibility of transfer failures."


puts "\nTransfer €2000 by an Agent".green
puts "\nBEFORE transfer"
puts bank1.name
puts account_1_1.holder + "´s account balance: €" + account_1_1.balance.to_s
puts "history of the transfers: " + bank1.transfers.count.to_s
puts bank2.name
puts account_2_1.holder + "´s account balance: €"  + account_2_1.balance.to_s
puts "history of the transfers: " + bank2.transfers.count.to_s

status = agent.makeTransfer(20000,account_1_1,account_2_1)

puts "\nTransfering...".cyan

puts "\nAFTER transfer"
puts "Transfer status: " + status
puts bank1.name
puts account_1_1.holder + "´s account balance: €" + account_1_1.balance.to_s
puts "history of the transfers: " + bank1.transfers.count.to_s
puts bank2.name
puts account_2_1.holder + "´s account balance: €" + account_2_1.balance.to_s
puts "history of the transfers: " + bank2.transfers.count.to_s
